### PR TITLE
Bugfix/ex 5374 tagging e2e tests

### PR DESCRIPTION
**Pull request by javieri-empathy**
This is the 3rd attempt to fix the flaky tagging e2e tests. Hopefully, this is the last part of this great series.

There were multiple things wrong with these tests. The issues are much easier to see using FireFox for running the tests.

# Cancelled fetch tagging requests

As you may know, fetch does not guarantee that the request is queued when it is executed in the unload process. In our case, what was happening in these tests is that usually the tracking request that was performed while leaving the page due to a click on a banner, result... was cancelled and this raised an error of an unhandled rejected promise.

Fortunately, Cypress can also intercept `navigator.sendBeacon` requests. So now this is the API that we are using in the mocked adapter for tracking requests.

# Mocked adapter was not used in tests.

Previously we had to pass a `useMockedAdapter=true` when navigating so this was the adapter used. When navigating to certain pages like the fake result ones this parameter was not passed. So in the add to cart tests for example, the real adapter was used. Now It is enough to load the demo view with Cypress.

# Cypress is too fast

I think this is the key of the issue of the randomly failing e2e tests, at least in Firefox. When asking Cypress to click a banner, result, promoted, or redirection, it did that immediately, as soon as the component was mounted to the DOM. Some of the siblings of that component weren't even yet mounted when this happened. But that is not really a problem. 

The problem was that there was a queue of asynchronous emitter events pending to be executed at the time the entity was clicked. These events sometimes included the `SearchTaggingChanged`, which is used by the `tagging` module to copy the query tagging with a defined debounce. As this event was not processed yet, the user click event (i.e. `UserClickedARedirection`), which is a synchronous event, was executed before it, and as a result it couldn't force to copy the query tagging.

TL;DR; Sometimes `UserClickedXXX` was executed before `SearchTaggingChanged` because of asynchronicity. This is (sadly) solved by adding an arbitrary wait. I couldn't find any other condition or guard to make cypress slowdown the click.

# Chunks

There was a warning that I saw a while ago complaining about the chunks order. Now I've removed unused views from the "production" environment, which is only used by Cypress. This should help us speed up our test server a little bit.

### CODE REVIEW

File: packages/x-components/src/utils/object.ts [hunk 1]
```diff
@@ -138,3 +138,29 @@ export function getNewAndUpdatedKeys<ObjectType extends Dictionary>(
...
+/**
+ * Ensures that the given condition is met in all the entries of the object.
+ *
+ * @param object - The object to check if every item meets the given condition.
+ * @param condition - The condition to check in each one of the entries of the object.
+ *
+ * @returns True when all the entries pass the condition. False otherwise.
+ * @public
+ */
+export function every<ObjectType extends Dictionary>(
```
Comment (javieri-empathy) [inline packages/x-components/src/utils/object.ts:151]: Initially I had manually stubbed the navigator `sendBeacon` method, and then used this `every` method to perform checks in the tests. Now this is not needed anymore, but I think this function can help us in the future.

File: packages/x-components/tests/e2e/cucumber/tagging/tagging.feature [hunk 2]
```diff
@@ -5,9 +5,7 @@ Feature: Tagging component
     And   a suggestions API
     And   a related tags API
     And   a recommendations API with a known response
-    And   a query tagging API
```
Comment (CachedaCodes) [inline packages/x-components/tests/e2e/cucumber/tagging/tagging.feature:8]: These 3 steps you removed were not used anywhere else, so I think their logic can be removed from `mocked-responses.spec.ts` (unless they might be used in the future instead of the new `a tracking API with a known response`).

File: packages/x-components/tests/e2e/cucumber/tagging/tagging.spec.ts [hunk 3]
```diff
@@ -1,22 +1,28 @@
 import { Given, Then, When } from 'cypress-cucumber-preprocessor/steps';
 
-Given('a URL with query parameter {string}', (query: string) => {
-  cy.visit(`/?useMockedAdapter=true&q=${query}`);
+Given('a tracking API with a known response', () => {
```
Comment (CachedaCodes) [inline packages/x-components/tests/e2e/cucumber/tagging/tagging.spec.ts:3]: Would it make sense to move this to `mocked-responses.spec.ts` with the other mocked tracking

**Review by CachedaCodes — CHANGES_REQUESTED** [review]

File: packages/x-components/src/utils/object.ts (hunk 1 above)
Comment (LuisMartinez15) [reply packages/x-components/src/utils/object.ts:151]: Hey bro! You're human! Welcome to the human realm! :trollface:

In a more serious note, we are fine with the addition of this utility. However, it currently has a bug. The `paramsMatch` flag "accumulator" is not working as intended, as it is being reset each iteration. We should not invert its value and we should use an `&&` operator instead of the `||` one, so the previous value is always taken into account. For example, in case in a previous iteration `paramsMatch` resulted in a `false` value, we shouldn't even care about the `condition` of the current/next ones, as at least one object key/entry has already failed the condition.

You can clearly see this behaviour happening in the second test you made, if you invert the order in the object, { a: '1', b: 2 }, you will see that the test fails, where it shouldn't, as we are only checking, in the end, that the last entry/key fulfils the condition.

As a side note, we could even optimise this method so it breaks the reduce execution whenever the condition is not met (probably not worth it for the majority of use cases, tho).

File: packages/x-components/tests/e2e/cucumber/filters/sliced-filters.spec.ts [hunk 4]
```diff
@@ -3,7 +3,7 @@ import { And, Given, Then, When } from 'cypress-cucumber-preprocessor/steps';
 let totalFilters: number, hiddenFilters: number;
 
 Given('following config: max of sliced filters {int}', (slicedFiltersMax: number) => {
-  cy.visit('/?useMockedAdapter=true');
+  cy.visit('/?');
```
Comment (LuisMartinez15) [inline packages/x-components/tests/e2e/cucumber/filters/sliced-filters.spec.ts:6]: I guess we can also remove the `?` here, right? Same for all the other files where you've applied this change (I'll just leave code suggestions in case you want to batch commit them instead of changing them manually):

```suggestion
  cy.visit('/');
```

File: packages/x-components/tests/e2e/cucumber/next-queries/next-queries.spec.ts [hunk 5]
```diff
@@ -21,7 +21,7 @@ Given(
       }
     };
 
-    cy.visit('/?useMockedAdapter=true', {
+    cy.visit('/?', {
```
Comment (LuisMartinez15) [inline packages/x-components/tests/e2e/cucumber/next-queries/next-queries.spec.ts:24]:
```suggestion
    cy.visit('/', {
```

File: packages/x-components/tests/e2e/cucumber/popular-searches/popular-searches.spec.ts [hunk 6]
```diff
@@ -18,7 +18,7 @@ Given(
         }
       }
     };
-    cy.visit('/?useMockedAdapter=true', {
+    cy.visit('/?', {
```
Comment (LuisMartinez15) [inline packages/x-components/tests/e2e/cucumber/popular-searches/popular-searches.spec.ts:21]:
```suggestion
    cy.visit('/', {
```

File: packages/x-components/tests/e2e/cucumber/query-suggestions/query-suggestions.spec.ts [hunk 7]
```diff
@@ -18,7 +18,7 @@ Given(
         }
       }
     };
-    cy.visit('/?useMockedAdapter=true', {
+    cy.visit('/?', {
```
Comment (LuisMartinez15) [inline packages/x-components/tests/e2e/cucumber/query-suggestions/query-suggestions.spec.ts:21]:
```suggestion
    cy.visit('/', {
```

File: packages/x-components/tests/e2e/cucumber/recommendations/recommendations.spec.ts [hunk 8]
```diff
@@ -11,7 +11,7 @@ Given('following config: max items to store is {int}', (maxItemsToRequest: numbe
     }
   };
 
-  cy.visit('/?useMockedAdapter=true', {
+  cy.visit('/?', {
```
Comment (LuisMartinez15) [inline packages/x-components/tests/e2e/cucumber/recommendations/recommendations.spec.ts:14]:
```suggestion
  cy.visit('/', {
```

File: packages/x-components/tests/e2e/cucumber/related-tags/related-tags.spec.ts [hunk 9]
```diff
@@ -16,7 +16,7 @@ Given(
       }
     };
 
-    cy.visit('/?useMockedAdapter=true', {
+    cy.visit('/?', {
```
Comment (LuisMartinez15) [inline packages/x-components/tests/e2e/cucumber/related-tags/related-tags.spec.ts:19]:
```suggestion
    cy.visit('/', {
```

File: packages/x-components/tests/e2e/cucumber/search-box/search-box.spec.ts [hunk 10]
```diff
@@ -19,7 +19,7 @@ Given(
         }
       }
     };
-    cy.visit('/?useMockedAdapter=true', {
+    cy.visit('/?', {
```
Comment (LuisMartinez15) [inline packages/x-components/tests/e2e/cucumber/search-box/search-box.spec.ts:22]:
```suggestion
    cy.visit('/', {
```

File: packages/x-components/tests/e2e/cucumber/tagging/tagging.spec.ts [hunk 11]
```diff
@@ -29,15 +35,15 @@ Then('result click tagging request is triggered', () => {
 });
 
 Then('query tagging request should be triggered', () => {
-  cy.wait('@queryTagging');
+  cy.wait('@queryTagging').should('exist');
 });
 
 Then('query tagging request has been triggered', () => {
-  cy.get('@queryTagging').should('exist');
+  cy.get('@queryTagging', { timeout: 0 }).should('exist');
```
Comment (LuisMartinez15) [inline packages/x-components/tests/e2e/cucumber/tagging/tagging.spec.ts:42]: Sadly, I'm afraid to say this is causing some tests to fail locally. And they keep failing randomly (sometimes they pass sometimes they don't 🤷🏿‍♂️ ).

image

Both tests failing in this image is due to same error in this line.

**Review by LuisMartinez15 — CHANGES_REQUESTED** [review]

File: packages/x-components/tests/e2e/cucumber/filters/sliced-filters.spec.ts (hunk 4 above)
Comment (javieri-empathy) [reply packages/x-components/tests/e2e/cucumber/filters/sliced-filters.spec.ts:6]: Global replaces in the morning do not work well

File: packages/x-components/src/utils/object.ts [hunk 12]
```diff
@@ -138,3 +138,29 @@ export function getNewAndUpdatedKeys<ObjectType extends Dictionary>(
...
+    key: keyof ObjectType,
+    value: Exclude<ObjectType[keyof ObjectType], undefined>,
+    index: number
+  ) => boolean
+): boolean {
+  return reduce(
+    object,
+    (paramsMatch, key, value, index) => {
+      return !paramsMatch || condition(key, value, index);
+    },
```
[the remark this answers is not in the document]
Comment (javieri-empathy) [reply packages/x-components/src/utils/object.ts:163]: You are right, I tried to go clever in this one and failed miserably. I changed the approach to use `Object.entries(object).every`, and we will gain that early return as soon as one condition fails.

File: packages/x-components/tests/e2e/cucumber/tagging/tagging.spec.ts (hunk 11 above)
Comment (LuisMartinez15) [reply packages/x-components/tests/e2e/cucumber/tagging/tagging.spec.ts:36]: As we talked in private, this seems to be working as expected (something might have been wrong in my local environment by the time I ran the tests).

**Review by mavmaf — COMMENTED** [review]
No issues seen after running it several times locally with different throttling and browsers.
Just one favor: it seems that at some point cypress.json file was merged by mistake. Please modify line #11 to the following:
`  "testFiles": ["cucumber/**/*.features"],`

**Comment by LuisMartinez15** [comment]
> No issues seen after running it several times locally with different throttling and browsers. Just one favor: it seems that at some point cypress.json file was merged by mistake. Please modify line #11 to the following: ` "testFiles": ["cucumber/**/*.features"],`

I think you had it changed locally and checked out this pull request (resulting in a merge with its content changed to your local modifications), because I see the `cypress.json` file with the same content you're asking to apply. This file hasn't been changed in pull request 😅

**Comment by mavmaf** [comment]
> > No issues seen after running it several times locally with different throttling and browsers. Just one favor: it seems that at some point cypress.json file was merged by mistake. Please modify line #11 to the following: ` "testFiles": ["cucumber/**/*.features"],`
> 
> I think you had it changed locally and checked out this pull request (resulting in a merge with its content changed to your local modifications), because I see the `cypress.json` file with the same content you're asking to apply. This file hasn't been changed in pull request 😅

Hahahaha F, what a _bocachancla_

**Review by CachedaCodes — APPROVED** [review]

**Review by LuisMartinez15 — APPROVED** [review]
LGTM

Good job!
